### ISSUE TITLE
MetalLB cannot peer with BGP routers that Calico is already peering with

### ISSUE THREAD

**Issue by WillieWookiee** [metallb/metallb#114]
**Is this a bug report or a feature request?**:

Question actually.

**What happened**:
Can't get MetalLB to peer with my core router.

**What you expected to happen**:
Peering is expected to happen and for me to see the routes for each node in the routing table.

**How to reproduce it (as minimally and precisely as possible)**:
Setup MetalLB and peer it with a Cisco L3 routing device.

**Anything else we need to know?**:
I am not sure if this is something related to the Cisco side or the MetalLB side. I also have calico peering with the same Cisco device with the same IP address and that could be the problem, but I wanted to verify. I am not sure that it is a bug.

Getting this in the log:
{"log":"E1213 22:09:35.960710       1 bgp.go:48] read OPEN from \"10.1.105.1:179\": message type is not OPEN, got 3, want 1\n","stream":"stderr","time":"2017-12-13T22:09:35.961076973Z"}

Makes me think the connection to calico is hijacking the metallb connection.

**Environment**:
- MetalLB version: Not sure how to find this.
- Kubernetes version: Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.5"
- BGP router type/version: Cisco 4500 Version 03.09.00.E
- OS (e.g. from /etc/os-release): "16.04.3 LTS (Xenial Xerus)"
- Kernel (e.g. `uname -a`): Linux 4.4.0-103-generic

**Comment by danderson** [comment]
Thank you very much for the report! As you'll see in this message, this is a rich bug report, containing at least 4 separate bugs/improvements :)

So, the error you're getting is that the peer router is sending MetalLB a BGP NOTIFICATION message ("oops, there's a problem, I'm closing the connection") instead of BGP OPEN ("Hi, I'm a router and here's my capabilities").

## Ease of use/debugging bugs

This points to a first bug: MetalLB does not decode the full notification, it aborts parsing as soon as it sees that the message is not the one it expected. This means we throw away the debugging information that the peer is sending us. Filed #115 to fix that.

Second, you said that you're not sure how to find the MetalLB version you're using. Good point! If you're running a released version, you can look at the container image version (with e.g. `kubectl get po -n metallb-system -l app=controller -oyaml`). Filed #116 to have MetalLB log its version on startup, and to update the issue template.

## The peering bug

Now, about your actual issue... I think there's probably 2 separate things going on here.

### Conflict with Calico

As you said, Calico is probably going to be an issue. AFAIK, BGP only allows 1 session between a pair of IPs. So if Calico is already peering, MetalLB cannot also peer with the router.

I have a couple of potential answers to what we can do about that. Some are policy/documentation (peer calico with your "datacenter fabric" routers, and peer MetalLB with your "datacenter edge" routers), another might be a technical solution (make MetalLB peer with the calico BGP daemon on the node, and inject routes that way). I've filed #117 to investigate that more.

### What error is the Cisco router sending?

The failure mode we would see with a Calico conflict doesn't seem to match what you're seeing. BGP has a resolution algorithm for conflicting BGP sessions (it has to, because the peering logic is not race-free, so it's common for routers to temporarily end up with >1 connection to a peer), but that algorithm iirc involves ungracefully closing the BGP session, not sending a NOTIFY. In fact, there doesn't seem to be any notification code that means "sorry, I have another BGP session for you already". So... The notification may be a separate interop issue with Cisco IOS specifically.

To confirm this, we could do one of two things:

- Are you comfortable with grabbing and sharing packet captures? If so, could you share a capture of tcp port 179 traffic, so that I can examine the BGP traffic and see what the error is? On your k8s nodes, the command for that would be `tcpdump -i any -w dump.pcap tcp port 179`. This will capture both Calico and MetalLB BGP traffic. Leave the capture running for at least 1min, to make sure it captures some connection attempts. Attach the pcap to this issue, and I'll examine it.
- Alternatively, later today I can prepare a debugging container image for MetalLB, with better logging for BGP notification messages. You can deploy that test image to your cluster, and grab the debugging data we need directly from MetalLB's logs.

**Comment by WillieWookiee** [comment]
Dave, thank you very much for the thorough response, it really helps.

Instead of going as far as tcpdumps (which I will be happy to do if needed) I did the debug on the Cisco side and believe I have confirmed what we suspected that the peers were colliding. 

I get the following message:

Dec 19 18:24:51.528: %BGP-3-NOTIFICATION: sent to neighbor 10.1.105.65 passive 6/7 (Connection Collision Resolution) 0 bytes
Dec 19 18:24:51.529: %BGP-5-NBR_RESET: Neighbor 10.1.105.65 passive reset (Peer closed the session)
Dec 19 18:24:51.529: %BGP-5-ADJCHANGE: neighbor 10.1.105.65 passive Down Error during connection collision

A quick google search showed that to be the case:

https://www.cisco.com/c/en/us/support/docs/ip/border-gateway-protocol-bgp/5816-bgpfaq-5816.html#anc40

**Comment by danderson** [comment]
Aha, perfect! Yes, that debug log confirms our suspicion about session collisions. The reason I was confused is that this code/subcode is defined in [RFC 4486](https://tools.ietf.org/html/rfc4486), and not the base BGP spec [RFC 4271](https://tools.ietf.org/html/rfc4271), so I had never seen the BGP Cease errors before.

Okay, so that fully explains what you're seeing, and confirms that Calico and MetalLB are conflicting with each other. Unfortunately, there's no obvious easy fix for that :(. The closest I can offer is something that is very specific to your network architecture: is there another router that you could peer MetalLB with, that is not also a Calico peer?

For example, in my cluster layout, I have top of rack BGP routers, and they connect to a pair of core routers that connect to the rest of the world. In that setup, I would peer Calico with the ToRs (and have them propagate their routes to the core), and I would peer MetalLB with the core routers. The reasoning is that the Calico session are distributing routes internally to the cluster's network, so it should be peering with the first hop outside the machine. OTOH, MetalLB wants to attract traffic from outside the cluster, so it should peer with the "border" of the cluster, i.e. the core routers that connect to the rest of the world.

Another alternative, if your router supports it, might be to define some VRFs and do some hacky cross-VRF route propagation, so that there are 2 logical routers, and Calico and MetalLB are not peering with the "same" router.

But, both of these are just hacky workarounds :(. I will investigate options for Calico interop, and we'll see if we can come up with something better. Worst case, we need to at least document this incompatibility, since it's a pretty big deal... But I'm hoping that I can find a way to make Calico and MetalLB coexists instead.

I'm duping #117 to this issue, and I'll keep using this issue to track the investigation on calico compatibility. Pasting the bug text from #117 below...

## MetalLB vs. Calico interop problems

Calico can be configured to peer with BGP routers, so that pod traffic routing between L3 network domains works. However, this puts Calico in conflict with MetalLB, because there can only be one node<>router BGP session, and Calico is consuming it.

We need at the very least some documentation about this:

- State that MetalLB is not compatible with Calico when you're using Calico's external peering feature
- Alternatively, provide some architectural guidance on how to peer both: Calico should peer with top-of-rack switches to distribute pod network routing data, and MetalLB should peer with the "datacenter edge" routers that are the interface between the cluster and the world. Requires a bunch of network diagrams and explanations to get across, probably a good candidate for an installation sub-article on the new website.

Separately, we should also investigate a technical solution. Is there a way to make MetalLB piggyback on Calico's BGP sessions? Can we somehow inject routes into the local Calico BGP speaker? If so, we could implement a new "calico" peer type, and teach MetalLB that for this peer type, it should talk to the local Calico daemon and inject its routes that way.

**Comment by danderson** [comment]
I will investigate in more detail later today after work, but after a _very_ quick reading of calico's documentation, one potential hack suggests itself...

Calico supports per-node BGP peering configurations. Assuming we can get the configuration to be acceptable in terms of the BGP spec, we could make MetalLB listen for BGP on a static host port (not 179), and create per-node BGP peerings in Calico to peer with MetalLB. Basically, make calico's bgpd on each node peer with localhost:1234, so that MetalLB can inject its routes into Calico that way.

Open questions:

- BGP peering with localhost is notoriously tricky, because the router can incorrectly believe that it's peering with itself. Calico uses the GoBGP codebase, which should not have this problem... But it needs to be tested.
- Does the Calico node daemon run the full BGP convergence algorithm? IOW, if we have a peering chain of MetalLB<>Calico<>external router, will Calico propagate the routes from MetalLB to the external router? In theory it should, because Calico advertises itself as "we just make your k8s nodes look like a regular BGP router", so unlike MetalLB it should be implementing the full BGP convergence/redistribution algorithm.
- Can we make this configuration automagic? Given appropriate RBAC rules, we could give MetalLB the permission to create new calico bgpPeer objects, so some component of MetalLB could automatically reconfigure the cluster to peer with MetalLB. This adds a bunch of complexity, and the "magic" may not be welcome by cluster/network admins who want explicit control over what happens to their network.
- Alternatively, is it possible to define the MetalLB peering as a "global" Calico BGP peer, with a peer address of 127.0.0.1? Again, in theory, this should just work: global peer will apply to all nodes, and all nodes will just connect to a different MetalLB instance on each machine. That should be fine... But maybe Calico has some sanity checks that prevent this. This would be much nicer from an admin perspective, because we can just tell Calico cluster operators "here's one BGP peer object that you should add to your Calico config, and voila, MetalLB just works!"

**Comment by WillieWookiee** [comment]
Thanks again for the detailed analysis. What about having MetalLB containers attain a different IP than the node and just let calico use the node's IP? This might take some configuration options on the cluster, but if the Cisco knew about MetalLB as a different IP, that would solve everything.

Unfortunately, I don't have another internal L3 device that I can peer with. I am already peering Calico with my core routers.

Another hacky method would be to introduce a BGP software router between MetalLB and the core router. Then it would be MetalLB<>Linux-basedBGPPeer<>Cisco core router.

Just some brainstorming here.

**Comment by danderson** [comment]
> What about having MetalLB containers attain a different IP than the node and just let calico use the node's IP?

Interesting thought! It would be possible, but would require a bunch of completely custom k8s node configuration (adding more IPs to the node). This makes MetalLB much harder to deploy, because suddenly choosing to use MetalLB has implications on how you provision your machines. It's definitely possible, but I'd like to keep it as a last resort, and try to have a solution that's more "zero config".

> Unfortunately, I don't have another internal L3 device that I can peer with. I am already peering Calico with my core routers.

Yeah, that's what I figured, and I suspect many Calico users are in your situation. Out of curiosity, what does this peering buy you? Does your core router redistribute these routes to ToRs? I'm trying to get a picture of your network topology, so I understand why you're exposing Calico's pod network to your physical network.

> Another hacky method would be to introduce a BGP software router between MetalLB and the core router. Then it would be MetalLB<>Linux-basedBGPPeer<>Cisco core router.

Interesting idea! I know at least one other MetalLB user who does something similar, they run a BIRD instance on each node, and peer it with both MetalLB and their upstream network. They're doing it for different reasons than Calico, but that model already works from MetalLB's perspective.

In a perfect world, I would still like to try and make MetalLB peer directly with Calico, so that we don't have to run yet-another-BGP router on the cluster (more CPU/memory overhead), but it's definitely an option. If direct Calico integration doesn't work out, that's probably the next best thing.

**Comment by WillieWookiee** [comment]
> Out of curiosity, what does this peering buy you? Does your core router redistribute these routes to ToRs?

We are exposing pods so that when we use a database solution that replicates, something like Cockroach DB, all of the instance of cockroach can communicate directly between each other. 
Maybe there is a different way to do this with K8 and I would be open to suggestions, but that kind of gets outside the scope of this issue.

**Comment by nyaxt** [comment]
We've successfully deployed MetalLB on k8s+calico cluster w/ a hacky workaround.
Instead of letting MetalLB join the iBGP directly, we pointed MetalLB speakers to a (separate) BGP route reflector, so that they would avoid having same IP addrs pairs in the peering graph.

**Comment by WillieWookiee** [comment]
> We've successfully deployed MetalLB on k8s+calico cluster w/ a hacky workaround.
Instead of letting MetalLB join the iBGP directly, we pointed MetalLB speakers to a (separate) BGP route reflector, so that they would avoid having same IP addrs pairs in the peering graph.

Yes, that's what I mentioned above as a possible work around. Good to know that it does work.

**Comment by danderson** [comment]
Romana is (somewhat) in the same boat: one of the configurations it supports uses BIRD route publishers to peer with the datacenter network and announce the cluster network: https://github.com/romana/romana/wiki/Romana-route-publisher

The addon supports providing custom BIRD config snippets, and I'm told by Romana folks that the route agent is configured such that it should redistribute advertisements just fine if metallb peers and injects routes... So, assuming BIRD is okay with localhost<>localhost peering, we should be golden there too.

**Comment by danderson** [comment]
Finally getting around to this...

Configured a Calico peer of 127.0.0.1, and told MetalLB to peer with 127.0.0.1... And it almost works! So far the only objection Calico has is that it's using the same router ID as MetalLB, so it thinks it's talking to itself.

Adding a config option to set MetalLB's router ID, let's see how this goes...

**Comment by danderson** [comment]
Making some progress. The peering with Calico is still pretty unstable right now because Calico is actively trying to connect, and therefore sometimes connects to itself and rejects the connection. This triggers error backoff, so it becomes increasingly difficult for metallb to successfully connect. Hopefully Calico has some way to specify custom target BGP ports, which would fix this.

Second problem: when the connection establishes, BIRD marks metallb-originated routes as unreachable, and so doesn't propagate them to other peers. It looks like it's marking the routes unreachable because the next-hop is an IP of a local interface (i.e. the node IP), and BIRD decides that this means the route is unreachable, for some reason...

**Comment by danderson** [comment]
Bad news, I think there's no way to make MetalLB cooperate with Calico in the way I imagined.

1. Calico does not support BGP peering options such as target port number or passive mode. This makes it impossible to avoid the race condition where calico connects to itself and enters error backoff, which means that the peering will be very unreliable. I can't fix this on MetalLB's side, Calico has to implement configuration options for BGP peering.
2. BIRD marks routes received with a next-hop of a local interface IP as unreachable. This _might_ not be a problem because Calico always sets next-hop-self when exporting, but it might mess up the local routing table. I would have investigated this more, but see the next point...
3. Calico filters out all routes that it doesn't control when exporting to peers. projectcalico/calicoctl#1138 theoretically addresses this by allowing the specification of custom BGP filters, but then projectcalico/calico#292 ruins that by saying they do not want to document the feature, and don't want to support it.

The real killer is the 3rd problem, but the 1st also makes peering with Calico really unreliable, sadly. So, there is no technical solution for making MetalLB work well with Calico :(

We can still make things work by implementing what @WillieWookiee suggested, and documenting how cluster operators can create an additional IP for each node, and use that IP for MetalLB peering. We need a small change to support setting the source IP in internal/bgp, but that's feasible at least.


Next up, I'm going to install Romana and see if that works any better.

**Comment by danderson** [comment]
Good news! Romana pretty much Just Works. You have to add the route publisher addon, and configure it just right, but when done right, MetalLB will peer with the local route publisher and redistribute routes to the upstream peers.

So, Romana support is just a question of cleaning the configs a bit and documenting how to set it up.

**Comment by danderson** [comment]
New documentation for how to run MetalLB and Romana together is at https://master--metallb.netlify.com/configuration/romana/ . It'll got to the live website in the next release.

As far as Calico is concerned... Sadly, right now all I can do is document the mediocre workarounds listed in this bug, we can't do anything as clean as Romana in the current state of the world. I'll file upstream bugs with Calico to document what we need from them, so hopefully we can do something better in the future.

**Comment by danderson** [comment]
Filed calicoproject/calico#1603 and calicoproject/calico#1604 to request the features we need to make Calico integration as smooth as Romana. Leaving this open, but marking that it's blocked on upstream work.

**Comment by WillieWookiee** [comment]
@danderson Thank you very much for all your work on this! I was not even aware of Romana. I will remove Calico and install Romana and test it with my K8 setup.

**Comment by chrismarino** [comment]
@WillieWookiee Ping me here or on Slack if you have any problems. Really would like to see this in action.

**Comment by Cytrian** [comment]
It's working for me now, with Calico version 3.0.1.

I've hacked the `/etc/calico/confd/templates/bird.cfg.template` and altered the
"Global peers" section, so BIRD is passive when the peer is 127.0.0.1. No ideal, but it's working.
I did this by just mounting my change into the Pod, I don't want to rebuild Calico now.
```
    # ------------- Global peers -------------
    [..]
    protocol bgp Global_{{$id}} from bgp_template {
      {{if eq $data.ip ("127.0.0.1")}}passive on; # Don't talk to yourself{{end}}
      neighbor {{$data.ip}} as {{$data.as_num}};
    }
    [..]
```

I added the MetalLB IP pool to calico (Create it with calicoctl). Then these IPs won't get filtered out. I don't yet know, what https://github.com/projectcalico/calico/issues/1604 would change here.
```
apiVersion: projectcalico.org/v3
kind: IPPoolList
items:
- apiVersion: projectcalico.org/v3
  kind: IPPool
  metadata:
    name: metallb-ip-pool
  spec:
    cidr: 192.168.10.0/24
    ipipMode: Always
    natOutgoing: false
```
When I now create a local peer for Calico on each Node, MetalLB is happily connecting to it.
It seems to work without a unique AS per Node or MetalLB instance.

On the Calico side:
```
apiVersion: projectcalico.org/v3
kind: BGPPeer
metadata:
  name: bgppeer-metallb
spec:
  asNumber: 64631
  peerIP: 127.0.0.1
```
and on the MetalLB side:
```
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: metallb-system
  name: config
data:
  config: |
    peers:
    - peer-address: 127.0.0.1
      peer-asn: 64622
      my-asn: 64631
    address-pools:
    - name: default
      protocol: bgp
      cidr:
      - 192.168.10.0/24
```

The setup is very similar to the Romana setup then.

**Comment by rikatz** [comment]
Don't know if this is applied here, but we can also propose to Calico guys a new Resource called 'route' that publishes a static route to some (or any) nodes inside the cluster.

This way MetalLB could talk directly with Calico (or using the Kubernetes CRD from Calico, when this is the backend being used) to announce a new IP route regardless of it's controlled or not by Calico.

And it would also improve Calico to be a 'bigger' router other than just for what it controls

**Comment by danderson** [comment]
So, speaking for MetalLB, I have a slight objection to that plan, because it opens a massive Pandora's box: having to add code (and tests) to individually support every network addon in a different way. I definitely don't have the resources to keep up with that, given how much the CNI ecosystem is exploding right now.

Right now, MetalLB sits in a handy position: either the CNI doesn't do BGP stuff and everything just works, or it does BGP stuff and, using only provider-specific _configuration_, you can make the two interoperate. This is the ideal state from my perspective as maintainer of something that has to interact with Calico.

Naturally, that's just my perspective, I'm sure there are arguments in the other direction as well :)

**Comment by rikatz** [comment]
You're right in your vision Dave, I was just wondering if this could be a solution :)

Anyway, the mentioned workaround with Route Reflector would work fine, but it adds a lot of complexity to a simple architecture.

By the other hand, using the solution mentioned [here](https://github.com/google/metallb/issues/114#issuecomment-363761032) would also be great but in this case we relies that there's always a '127.0.0.1' peer running, and this could not be the case unless we run metallb speaker on every single node as a DaemonSet. 

Anyway this could be also a workaround for Romana and so.

I'll keep thinking in some better solution for this, and keep also following this thread :)

**Comment by danderson** [comment]
Thanks for thinking about more solutions! It's definitely appreciated.

So far, I'm optimistic that the things we need will get implemented by the Calico folks, they seemed receptive to the bugs and I've heard that it's on the plate for the next release.

**Comment by jpiper** [comment]
@Cytrian Have you tried setting the disabled flag to true on the calico IPPool object ([see here](https://docs.projectcalico.org/master/reference/calicoctl/resources/ippool))? This would make it so that calico doesn't ever try to allocate IPs from this pools. Likewise, surely the `ipipMode: Always` part is redundant

**Comment by jpiper** [comment]
For people wanting the exact code to mount the modified `bird.cfg.template` into calico (I haven't tested this but it should work).

Use this in your Calico manifest to inject the new bird template

```yaml
containers:
- volumeMounts:
  - name: bird-template
    mountPath: /etc/calico/confd/templates/bird.cfg.template
    subPath: bird.cfg.template
volumes:
- name: bird-template
  configMap:
    name: calico-metallb-config
```

And here's an example of the v3.0.2 calico template with the patch

```yaml
# This ConfigMap is used to configure a self-hosted Calico installation.
kind: ConfigMap
apiVersion: v1
metadata:
  name: calico-metallb-config
data:
  bird.cfg.template: |-
    # Generated by confd
    include "bird_aggr.cfg";
    include "bird_ipam.cfg";
    {{$node_ip_key := printf "/host/%s/ip_addr_v4" (getenv "NODENAME")}}{{$node_ip := getv $node_ip_key}}

    router id {{$node_ip}};

    {{define "LOGGING"}}
    {{$node_logging_key := printf "/host/%s/loglevel" (getenv "NODENAME")}}{{if exists $node_logging_key}}{{$logging := getv $node_logging_key}}
    {{if eq $logging "debug"}}  debug all;{{else if ne $logging "none"}}  debug { states };{{end}}
    {{else if exists "/global/loglevel"}}{{$logging := getv "/global/loglevel"}}
    {{if eq $logging "debug"}}  debug all;{{else if ne $logging "none"}}  debug { states };{{end}}
    {{else}}  debug { states };{{end}}
    {{end}}

    # Configure synchronization between routing tables and kernel.
    protocol kernel {
      learn;             # Learn all alien routes from the kernel
      persist;           # Don't remove routes on bird shutdown
      scan time 2;       # Scan kernel routing table every 2 seconds
      import all;
      export filter calico_ipip; # Default is export none
      graceful restart;  # Turn on graceful restart to reduce potential flaps in
                         # routes when reloading BIRD configuration.  With a full
                         # automatic mesh, there is no way to prevent BGP from
                         # flapping since multiple nodes update their BGP
                         # configuration at the same time, GR is not guaranteed to
                         # work correctly in this scenario.
    }

    # Watch interface up/down events.
    protocol device {
      {{template "LOGGING"}}
      scan time 2;    # Scan interfaces every 2 seconds
    }

    protocol direct {
      {{template "LOGGING"}}
      interface -"cali*", "*"; # Exclude cali* but include everything else.
    }

    {{$node_as_key := printf "/host/%s/as_num" (getenv "NODENAME")}}
    # Template for all BGP clients
    template bgp bgp_template {
      {{template "LOGGING"}}
      description "Connection to BGP peer";
      local as {{if exists $node_as_key}}{{getv $node_as_key}}{{else}}{{getv "/global/as_num"}}{{end}};
      multihop;
      gateway recursive; # This should be the default, but just in case.
      import all;        # Import all routes, since we don't know what the upstream
                         # topology is and therefore have to trust the ToR/RR.
      export filter calico_pools;  # Only want to export routes for workloads.
      next hop self;     # Disable next hop processing and always advertise our
                         # local address as nexthop
      source address {{$node_ip}};  # The local address we use for the TCP connection
      add paths on;
      graceful restart;  # See comment in kernel section about graceful restart.
    }

    # ------------- Node-to-node mesh -------------
    {{if (json (getv "/global/node_mesh")).enabled}}
    {{range $host := lsdir "/host"}}
    {{$onode_as_key := printf "/host/%s/as_num" .}}
    {{$onode_ip_key := printf "/host/%s/ip_addr_v4" .}}{{if exists $onode_ip_key}}{{$onode_ip := getv $onode_ip_key}}
    {{$nums := split $onode_ip "."}}{{$id := join $nums "_"}}
    # For peer {{$onode_ip_key}}
    {{if eq $onode_ip ($node_ip) }}# Skipping ourselves ({{$node_ip}})
    {{else if ne "" $onode_ip}}protocol bgp Mesh_{{$id}} from bgp_template {
      neighbor {{$onode_ip}} as {{if exists $onode_as_key}}{{getv $onode_as_key}}{{else}}{{getv "/global/as_num"}}{{end}};
    }{{end}}{{end}}{{end}}
    {{else}}
    # Node-to-node mesh disabled
    {{end}}


    # ------------- Global peers -------------
    {{if ls "/global/peer_v4"}}
    {{range gets "/global/peer_v4/*"}}{{$data := json .Value}}
    {{$nums := split $data.ip "."}}{{$id := join $nums "_"}}
    # For peer {{.Key}}
    protocol bgp Global_{{$id}} from bgp_template {
      {{if eq $data.ip ("127.0.0.1")}}passive on; # Don't talk to yourself{{end}}
      neighbor {{$data.ip}} as {{$data.as_num}};
    }
    {{end}}
    {{else}}# No global peers configured.{{end}}


    # ------------- Node-specific peers -------------
    {{$node_peers_key := printf "/host/%s/peer_v4" (getenv "NODENAME")}}
    {{if ls $node_peers_key}}
    {{range gets (printf "%s/*" $node_peers_key)}}{{$data := json .Value}}
    {{$nums := split $data.ip "."}}{{$id := join $nums "_"}}
    # For peer {{.Key}}
    protocol bgp Node_{{$id}} from bgp_template {
      neighbor {{$data.ip}} as {{$data.as_num}};
    }
    {{end}}
    {{else}}# No node-specific peers configured.{{end}}
```

**Comment by jpiper** [comment]
Ok, I can confirm that the override above works without any issues.

All you need to do mount in that modified config, and then create a disabled IPPool in calico to get it to re-advertise the prefixes that metallb is pumping in.

```
cat << EOF | calicoctl create -f -
apiVersion: projectcalico.org/v3
kind: IPPool
metadata:
  name: metallb-ip-pool
spec:
  cidr: 10.110.0.0/16
  disabled: true
EOF
```

Then configure your Calico Peering:

```
cat << EOF | calicoctl create -f -
apiVersion: projectcalico.org/v3
kind: BGPPeer
metadata:
  name: metallb
spec:
  peerIP: 127.0.0.1
  asNumber: 65480
EOF
```

and Metallb (in this case, calico has an ASN of 65479)

```
    - peer-address: 127.0.0.1
      peer-asn: 65479
      my-asn: 65480
```

Note: This needs to be an eBGP session, I got a load of errors if the ASNs are the same.

@danderson if you like I can write this up as a docs PR for you if you like?

**Comment by hightoxicity** [comment]
Hi guys, combining a lot of those tricks I am able to make metallb announces a public ip near the calico's bird at layer 3. I created a service using externalTrafficPolicy: Local
But I have an ARP problem on the host with both calico and metal lb doing announces for distinct ips on same virtual interface

 ip n
194.187.170.11 dev cali80a929fc019  FAILED
192.168.56.115 dev ens192 lladdr 00:50:56:9f:24:c1 REACHABLE
172.20.0.1 dev ens224 lladdr 00:50:56:db:05:f4 STALE
172.22.89.104 dev cali80a929fc019 lladdr ae:b6:34:97:df:e6 PERMANENT
172.22.89.110 dev cali27acbdcf300 lladdr a6:a7:9b:40:ca:11 PERMANENT
192.168.56.254 dev ens192 lladdr 00:50:56:b7:5e:21 DELAY
172.20.0.2 dev ens224  FAILED


metallb is trying to announce 194.187.170.11 on the interface cali80a929fc019 when calico already do it for 172.22.89.104, an ip form pod networking cidr.

Am I the lonely poor guy to encounter this problem?

Thanks for your help.

**Comment by hightoxicity** [comment]
OK it seems I am exaclty in the following case:[https://github.com/google/metallb/issues/258](url)

**Comment by caseydavenport** [comment]
I think a desirable architecture here might be for Calico to advertise these IPs in place of the MetalLB `speaker` daemonset, so that you don't need to run two BGP agents on each node.

In that case, we'd update Calico to support advertising Service IPs, and then simply run the `controller` deployment.

WDPT?

**Comment by danderson** [comment]
There's precedent for that, kube-router has done effectively the same thing.

It seems like a reasonable architecture to me, with a couple of caveats:

 - Calico must implement `externalTrafficPolicy: {Local,Cluster}` correctly. In Cluster mode, all "healthy" nodes (where Calico gets to decide what healthy means) must advertise the service IP. In Local mode, only "healthy" nodes with _local_ pods for the service should advertise, because other nodes will blackhole any traffic you send them. This means the node agent must track the service's Endpoints object and know if the service is currently being hosted on the local node or not.
 - If you support BGP multipath, in Local mode you should ideally advertise 1 path for each local pod, so that the upstream routers do weighted LB depending on the number of pods. MetalLB doesn't do this today, just a suggestion if you have a way to do that easily.
 - This might cause some confusion because various bits of MetalLB's config will become no-ops. We'll have to carefully document that stuff like setting custom BGP communities or generating aggregate announcements don't apply if you're not using MetalLB's speaker.
 - By removing the speaker from the equation, you'll miss out on future planned improvements to MetalLB's behavior - most notably, active healthchecking of endpoints as a condition to advertising. IOW, it's easy to get parity with MetalLB today, but removing advertisement control from MetalLB may interact poorly with future features.
 - The implication of the above is that this setup may stop working in some future version of MetalLB. I'll do my best to not gratuitously break it of course, but if some future major feature absolutely requires that I have control over the announcements... Then we have a problem, unless you can also offer me some mechanism to control the announcements that Calico is making (i.e. I have some place where I can say "actually don't announce this IP right now" and have the Calico node agent obey).
 - I'm happy to document this setup as a setup that should work as of today, but I can't really offer support for this kind of hybrid setup, because I don't know Calico's internals well enough to debug (not to mention that it's just me, in my spare time, and I have other hobbies :) ). That means I'll be sending people your way when they have a problem with this setup, once I've established that MetalLB is allocating IPs correctly, and from there it's your problem. I assume you're okay with that because unlike me you have the resources to handle support :).

But again, overall, seems like a plausible architecture, and kube-router has de-facto done this already, so I can document it as a "class" of network addons that do their own thing. What are your thoughts about my thoughts? :)

**Comment by caseydavenport** [comment]
@danderson sorry for the long delay here.

Yes, those are all very sensible considerations. And, I totally agree you shouldn't be on the hook should something go wrong with the Calico integration.

Some thoughts:
- Yep, Calico is already aware of `externalTrafficPolicy` for other advertisement features, so this should  be an easy one.
- BGP multipath - that's a good idea, will do some investigation into how feasible it is :)

I will need to think on this a bit more. I feel like there is value in simplifying the architecture and removing the need for _two_ BGP speakers on the node, but duplicating effort isn't really desirable, or easily maintainable in the long run.

Perhaps we simply just need to implement and support both modes on the Calico side, so people can pick and choose based on what they need more.

**Comment by danderson** [comment]
Another thought that I think I've voice in the past: if the Calico node agent gave me a way to peer with it over localhost and inject routes, that would be just a documentation change to MetalLB, similar to the currently documented approach with the Romana networking stack.

Of course, that creates more work for you in several dimensions, because you now need to propagate routes learned from localhost to external peers (but probably not to the rest of the calico-node mesh, if present), and possibly do some policy validation that I'm not stomping on the pod network. Those are both relatively easy to do in Bird, if you wanted to go that route I'd be happy to help come up with the configs.

Last time I suggested this it was rejected because it opened up too big of a can of worms, but Calico's changed a lot since then, so maybe it makes more sense today?

In my perfect world, I wouldn't have to live in a universe where MetalLB's behavior varies drastically based on what network addon is installed. To some extent it's a losing battle because network addon developers similarly want to own LB networking, so there's a natural conflict there.

**Comment by caseydavenport** [comment]
> Another thought that I think I've voice in the past: if the Calico node agent gave me a way to peer with it over localhost and inject routes, that would be just a documentation change to MetalLB

So, @danderson, I was working with @stevegaossou today trying to reproduce the original issues you hit while attempting this, and we didn't see the same behaviors mentioned - Calico trying to peer with itself and hitting the backoff. In fact, we saw Calico peer with MetalLB seamlessly over 127.0.0.1 without needing passive set at all on the peering.

I'm a bit suspicious of the result, since clearly lots of folks have hit some variation of this, but wondering if there might be something we did differently that made it work... The main thought I had is that Calico is selecting its BGP source address as the primary IP of the node, which is different from the configured MetalLB peer on localhost.

The steps we took were dead simple, much like others have described above but without needing passive set. We started with the MetalLB minikube example and then:
- Configured Calico to peer with MetalLB over localhost using a BGPPeer
- Configured MetalLB to peer with Calico over localhost
- BGP session goes to Established
- Configured an IPPool in Calico with the IP range given to MetalLB (to enable route export).

Do you recall what differences might have existed in your test environment that might have caused this difference in result?

**Comment by danderson** [comment]
Interesting!

First off, I haven't tried this configuration myself in about a year. I can't think of anything that would change the behavior significantly, but that's a point to consider.

The repro steps you described are pretty much exactly what I did. I will note that peering was intermittently successful when I tried: Calico was actively trying to connect to localhost (i.e. itself) and then disconnecting when it correctly detected the loop. If MetalLB slipped in at the right time between attempts and at the right spot in the backoff cycle, it was able to peer and break the cycle. However, it was a rare occurence on my 3-node cluster, meaning that at best 1 in 3 nodes successfully peered, and the others started spinning on hitting the backoff timers on the calico side.

One possible change to your repro to "help" with this: after Calico is configured with a localhost BGPPeer, let it sit like that for ~5min so that it can attempt+fail to connect a bunch of times. Then configure MetalLB and see if the peering works out.

If that doesn't help, once MetalLB is Established, try removing the MetalLB speaker DaemonSet and recreating it after 5min. In both cases, the thing I'm trying to do is to give calico's BGPd enough time to get into a backed-off state trying to reestablish the session, if that makes sense?


Your hypothesis on source IPs might be implicated... Specifically, you say you select the primary node IP as the BGP source address. Do you also configure Calico to only listen on specific addresses? Possibly the change could be that Calico is no longer listening on 127.0.0.1:179, so when MetalLB isn't around the failure mode is nice and clean (connection refused, not "hey you have a BGP loop misconfiguration").

Those are my thoughts so far. I'm travelling for work this week and next with minimal time to dig further, but when I get back I can try setting up a test case in virtuakube (metallb's shiny new e2e testing setup) and see if I can give you a crisp reproduction recipe. I already have k8s+calico in the test matrix, it's just a question of adding a test that exercises this peering config.

**Comment by caseydavenport** [comment]
Yes, looking further I think my hypothesis was perhaps misguided. We set `source address` but we don't set `strict bind on`, so I think BIRD should still be listening on all interfaces. 

We actually do see what appear to be connection attempts from ourselves, but they appear to be coming from a different address on the machine rather than the address Calico has chosen to use as its `source address`, so they simply show up as "connect from unexpected peer" rather than a loop misconfiguration, which might be the key difference here. However, even changing so that the incoming connections are from the same IP as the `source address` provided to BIRD, we're not seeing any intermittent connections or back-off behavior, and MetalLB establishes right away. 

> One possible change to your repro to "help" with this: after Calico is configured with a localhost BGPPeer, let it sit like that for ~5min so that it can attempt+fail to connect a bunch of times. Then configure MetalLB and see if the peering works out.

yep, thought this might be the case too but can't seem to repro. 

> Those are my thoughts so far. I'm travelling for work this week and next with minimal time to dig further, but when I get back I can try setting up a test case in virtuakube (metallb's shiny new e2e testing setup) and see if I can give you a crisp reproduction recipe. I already have k8s+calico in the test matrix, it's just a question of adding a test that exercises this peering config.

Looking forward to this - would be really nice to get to the bottom of this and document a working configuration. If we can manage to repro, then we're happy to implement the passive mode configuration, but just trying to make sure it's required first :)

**Comment by hightoxicity** [comment]
> > Another thought that I think I've voice in the past: if the Calico node agent gave me a way to peer with it over localhost and inject routes, that would be just a documentation change to MetalLB
> 
> So, @danderson, I was working with @stevegaossou today trying to reproduce the original issues you hit while attempting this, and we didn't see the same behaviors mentioned - Calico trying to peer with itself and hitting the backoff. In fact, we saw Calico peer with MetalLB seamlessly over 127.0.0.1 without needing passive set at all on the peering.
> 
> I'm a bit suspicious of the result, since clearly lots of folks have hit some variation of this, but wondering if there might be something we did differently that made it work... The main thought I had is that Calico is selecting its BGP source address as the primary IP of the node, which is different from the configured MetalLB peer on localhost.
> 
> The steps we took were dead simple, much like others have described above but without needing passive set. We started with the MetalLB minikube example and then:
> 
> * Configured Calico to peer with MetalLB over localhost using a BGPPeer
> * Configured MetalLB to peer with Calico over localhost
> * BGP session goes to Established
> * Configured an IPPool in Calico with the IP range given to MetalLB (to enable route export).
> 
> Do you recall what differences might have existed in your test environment that might have caused this difference in result?

I followed exactly what you suggested, and it is is working, my external IP can be reached, but the annoying part is that calico is reporting connection issues and another strange thing is that the prefix imported into bird are marked unreachable even if it works...

I fixed those two issues doing the following:
- patching the bird template and adding **passive on** into the peer config for 127.0.0.1
- Since I was using a distinct ASN at metallb side (I have an ibgp between my TOR and the node), I added a **local as METALLB_ASN** in the peer config, the routes are no more unreachable.

**Comment by falfaro** [comment]
On my test environment, I am trying to have MetalLB in BGP mode on a Charmed Distribution of Kubernetes (Canonical Kubernetes) using Calico as the CNI.

The problem that I am facing is that I need an external BGP entity (pfSense) that MetalLB peers, and have the traffic to LoadBalancer IPs flow via pfSense. Else, Kubernetes nodes or pods can't reach LoadBalancer IPs. This seems very inefficient to me: a Pod sends traffic to a LoadBalancer IP via pfSense and from pfSense back to the Kubernetes cluster. What I am seeking here is to have MetalLB peer with Calico, and then have Calico peer with pfSense to advertise the LoadBalancer IPs to the external world.

My understand is that @hightoxicity and @caseydavenport and @stevegaossou are working on having MetalLB peer with Calico. Any progress on this? On having this set up validated and documented?

**Comment by caseydavenport** [comment]
@falfaro This isn't something that I've been actively working on recently. I think Calico is missing the ability to set `passive on` in the BGP peer resource.

This is the issue tracking that: https://github.com/projectcalico/calico/issues/1603 (there's a comment explaining the work that needs to be done, and even a half-finished PR that has since been closed). It should be fairly straightforward to do, if anyone has the appetite.

**Comment by iceman91176** [comment]
Hello guys,

using a cusomized bird.cfg template solves the issues with peerings from localhost. Combined with disabled adress-pools calico is advertising the service-ips just fine. 

But using 127.0.0.1 has other implications. Consider an iBGP-Peer between Calico and the ToR-Switch/Router which is fairly usual in an AS-per-Rack architecture. So calico announces the routes learned from metallb to the ToR-router with a next-hop-address of 127.0.0.1, since it is an iBGP-session.
The ToR-router/switches can never reach that address. Ii is an invalid next-hop and will be discarded. On our arista-based switches a single invalid-next-hop even leads to a tear-down of the bgp-session :-1:  

Our approach was to modify the export filter, in a way that i sets the next-hop to the real-ip adress of the calico-node as soon as it encounters  a next-hop of 127.0.0.1.
Since i wasn't able to get the calico-nodes IP from etcd (although it is working in the bird.cfg.template), we are simply using the CALICO_ROUTER_ID, which is set to the host-ip in our case anyway.

Basically it looks like this

```
apiVersion: v1
data:
  bird_ipam.cfg.template: |-
      # Generated by confd
      {{- $node_ip := getenv "CALICO_ROUTER_ID"}}

      function fix_next_hop ()
      {
        {{if eq "" ($node_ip) -}}
        # No Node-IP found in ENV CALICO_ROUTER_ID - cannot fix next-hop
        {{- else -}}
        if bgp_next_hop ~ [127.0.0.1] then {
          bgp_next_hop = {{ $node_ip}};
        } {{- end }}
      }

      filter calico_export_to_bgp_peers {
        calico_aggr();
        fix_next_hop();
        .... rest of "vanilla" bird_ipam.cfg.template
```` 

If anyone is interested - here is the complete Calico-Manifest (v3.11) with customized templates https://gist.github.com/iceman91176/80edd9a0d517d3a448139958d59969b7

Since we are already modifying the default export filter, it is a simple task to accept routes from metallb  based on a community. In that case no Dummy-IPPool has to be created.

**Comment by logan2211** [comment]
For now I built custom templates with a [number of fixes](https://github.com/projectcalico/confd/compare/master...logan2211:master) that make calico-node operate more smoothly in a bare metal environment with MetalLB.

Patched calico-node images are available at https://github.com/logan2211/calico-node-baremetal/tree/v3.12.0

The [kustomize](https://github.com/logan2211/calico-node-baremetal/tree/v3.12.0/kustomize) present in the repo allows you to announce IPs from MetalLB using community string `<calico asn>:100` that will push routes through from MetalLB to your node/global calico peers without requiring creation of calico ippool resources. MetalLB community configuration is documented [here](https://metallb.universe.tf/configuration/#advertisement-configuration).

**Comment by ivan046** [comment]
@logan2211 Thanks for your impressive work! Can't wait that to be merged into the next Calico release!

**Comment by ivan046** [comment]
@logan2211can you, please, provide an example of MetalLb bgp config that works with your calico version (BGP_COMMUNITY=100)?

**Comment by logan2211** [comment]
> @logan2211can you, please, provide an example of MetalLb bgp config that works with your calico version (BGP_COMMUNITY=100)?

Sure! Check out https://gist.github.com/logan2211/bd0d8c3fd9091ddb207cde57e16d73a4

**Comment by carpenike** [comment]
Very excited to see this move forward as well! Do the maintainers intend to merge this into this repo?

**Comment by gclawes** [comment]
I've been able to get this working without the BGP_COMMUNITY stuff:
https://gitlab.com/gclawes/pico-cluster/-/blob/master/helm/metallb-values.yaml#L1
https://gitlab.com/gclawes/pico-cluster/-/blob/master/calico/bgp-configuration.yaml

All I needed to do was add the MetalLB ip ranges as an `IPPool` object in calico, and set it to `disabled: true` so calico doesn't use it for pod assignments.

```
apiVersion: crd.projectcalico.org/v1
kind: IPPool
metadata:
  name: metallb-ippool
spec:
  cidr: 10.102.0.0/16
  disabled: true
```

**Comment by fasaxc** [comment]
Calico now has the native ability to advertise external IPs from services.  Would that do the trick here? 

* metallb populates the external IP
* Calico advertises the external IP
* metallb disables its BGP peering

https://docs.projectcalico.org/networking/advertise-service-ips#advertise-service-external-IP-addresses

**Comment by champtar** [comment]
@fasaxc be carreful, ExternalIPs != LoadBalancerIPs
Just use ExternalIPs and remove MetalLB entirely

**Comment by gclawes** [comment]
@champtar are you thinking of ExternalIPs for Nodes?  Those are different from the EXTERNAL-IP field that a LB assigns to a Service (which is really `.status.loadBalancer.ingress`):

```
$ kubectl -n kube-system get svc ingress-nginx-controller -o wide
NAME                       TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)                      AGE    SELECTOR
ingress-nginx-controller   LoadBalancer   10.101.77.219   10.10.5.192   80:32504/TCP,443:32722/TCP   3d8h   app.kubernetes.io/component=controller,app.kubernetes.io/instance=ingress-nginx,app.kubernetes.io/name=ingress-nginx
$ kubectl -n kube-system get svc ingress-nginx-controller -o yaml | grep -A4 ^status:
status:
  loadBalancer:
    ingress:
    - ip: 10.10.5.192
```

@fasaxc I think that just advertises an entire prefix given in `serviceExternalIPs`, from the testing I've done it doesn't selectively advertise LB IPs that have been assigned to services like MetalLB does.

**Comment by champtar** [comment]
@gclawes I'm talking about Service 'externalIPs' field (https://kubernetes.io/docs/concepts/services-networking/service/#external-ips)
EXTERNAL-IP can be either ExternalIPs or LoadBalancerIPs (to make it extra confusing)
I haven't looked at Calico code but the documentation talks about external IP and they don't talk about running any loadBalancer controller, so that make me think that they really mean external IP ;)

**Comment by ryholt** [comment]
Hmm, it does seem to indicate that if you configure `externalTrafficPolicy: local` that you can use LoadBalancerIPs?

> Services must be configured with the correct service type (“Cluster” or “Local”) for your implementation. For externalTrafficPolicy: Local, the service must be type LoadBalancer or NodePort.

**Comment by fasaxc** [comment]
That's what I thought but I didn't work directly on this feature (I am a core maintainer of Calico, though).

**Comment by champtar** [comment]
@ryholt indeed, I was just reading "Advertise service external IP addresses" part ...
But maybe everything already work with ExternalIPs, and you can run 1 less piece of code (MetalLB)
https://github.com/projectcalico/calico/issues/3443
https://github.com/projectcalico/calico/pull/2868#pullrequestreview-291422668

**Comment by ryholt** [comment]
@champtar -- ah good thought. Just register the ClusterIPs in DNS instead? One use case for `LoadBalancerIP` are when you need a static IP that doesn't change (DNS servers, NFS servers, etc).

**Comment by ryholt** [comment]
Sorry, I read the https://kubernetes.io/docs/concepts/services-networking/service/#external-ips referenced above. You could theoretically define any external IP you want on the service itself and have Calico advertise it as a /32.

**Comment by champtar** [comment]
> Sorry, I read the https://kubernetes.io/docs/concepts/services-networking/service/#external-ips referenced above. You could theoretically define any external IP you want on the service itself and have Calico advertise it as a /32.

Exactly !
LoadBalancer is automatic self service from a pool of ip, externalIP is good old manual static ip
And to be sure to confuse people, kubectl EXTERNAL-IP show both ...

They should be the same thing on the data plane but are implemented differently in kube-proxy, to be sure to duplicate bugs :P

**Comment by ryholt** [comment]
So as long as you don't need a pool of IPs that auto assign is there any other reason to use Metallb over externalIP?

**Comment by champtar** [comment]
I don't see any if it works ;)

**Comment by ryholt** [comment]
Well I'll be damned... it works. :)

BGPConfiguration
```
apiVersion: v1
items:
- apiVersion: crd.projectcalico.org/v1
  kind: BGPConfiguration
  metadata:
    annotations:
      projectcalico.org/metadata: '{"uid":"985f849b-fc4e-4514-b9d6-4acfb0e6d6d3","creationTimestamp":"2020-07-29T14:00:36Z"}'
    creationTimestamp: "2020-07-29T14:00:37Z"
    generation: 1
    name: default
    resourceVersion: "2720013"
    selfLink: /apis/crd.projectcalico.org/v1/bgpconfigurations/default
    uid: 654e14bf-7520-4a94-8862-b2a61d36c76b
  spec:
    serviceClusterIPs:
    - cidr: 10.43.0.0/16
    serviceExternalIPs:
    - cidr: 10.45.0.0/16
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

Service config:
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    helm.fluxcd.io/antecedent: default:helmrelease/plex
    metallb.universe.tf/allow-shared-ip: plex
  creationTimestamp: "2020-07-24T11:10:01Z"
  labels:
    app.kubernetes.io/instance: plex
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: plex
    app.kubernetes.io/version: 1.19.1.2645-ccb6eb67e
    helm.sh/chart: plex-1.6.1
  name: plex-tcp
  namespace: default
  resourceVersion: "2749251"
  selfLink: /api/v1/namespaces/default/services/plex-tcp
  uid: 187f8b59-cab4-4151-a7d0-df4a92733126
spec:
  clusterIP: 10.43.129.214
  externalIPs:
  - 10.45.100.100
  externalTrafficPolicy: Local
  healthCheckNodePort: 31466
  ports:
  - name: pms
    nodePort: 30499
    port: 32400
    protocol: TCP
    targetPort: pms
  - name: http
    nodePort: 32708
    port: 80
    protocol: TCP
    targetPort: pms
  - name: https
    nodePort: 32287
    port: 443
    protocol: TCP
    targetPort: pms
  - name: plex-dlna
    nodePort: 31848
    port: 1900
    protocol: TCP
    targetPort: plex-dlna
  selector:
    app.kubernetes.io/instance: plex
    app.kubernetes.io/name: plex
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - ip: 10.20.50.155
```

After initial deployment, routes updated as expected on router:

```
ryan@vyos:~$ netstat -rn
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
0.0.0.0         PROVIDER    0.0.0.0         UG        0 0          0 bond0.4000
10.9.18.0       0.0.0.0         255.255.255.0   U         0 0          0 bond0.1
10.20.0.0       0.0.0.0         255.255.0.0     U         0 0          0 bond0.20
10.30.0.0       0.0.0.0         255.255.0.0     U         0 0          0 bond0.30
10.40.0.0       0.0.0.0         255.255.255.0   U         0 0          0 wg1
10.43.0.0       10.20.10.10     255.255.0.0     UG        0 0          0 bond0.20
10.43.21.128    10.20.10.10     255.255.255.192 UG        0 0          0 bond0.20
10.43.67.128    10.20.10.15     255.255.255.192 UG        0 0          0 bond0.20
10.43.74.128    10.20.10.17     255.255.255.192 UG        0 0          0 bond0.20
10.43.94.192    10.20.10.11     255.255.255.192 UG        0 0          0 bond0.20
10.43.106.163   10.20.10.10     255.255.255.255 UGH       0 0          0 bond0.20
10.43.117.90    10.20.10.12     255.255.255.255 UGH       0 0          0 bond0.20
10.43.129.214   10.20.10.10     255.255.255.255 UGH       0 0          0 bond0.20
10.43.144.128   10.20.10.12     255.255.255.192 UG        0 0          0 bond0.20
10.43.192.64    10.20.10.16     255.255.255.192 UG        0 0          0 bond0.20
10.45.0.0       10.20.10.10     255.255.0.0     UG        0 0          0 bond0.20
10.45.100.100   10.20.10.10     255.255.255.255 UGH       0 0          0 bond0.20
```

Cordoned the node, deleted the pod, plex starts up on other node, routes propagated as expected in router:

```
ryan@vyos:~$ netstat -rn
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
0.0.0.0         PROVIDER    0.0.0.0         UG        0 0          0 bond0.4000
10.9.18.0       0.0.0.0         255.255.255.0   U         0 0          0 bond0.1
10.20.0.0       0.0.0.0         255.255.0.0     U         0 0          0 bond0.20
10.30.0.0       0.0.0.0         255.255.0.0     U         0 0          0 bond0.30
10.40.0.0       0.0.0.0         255.255.255.0   U         0 0          0 wg1
10.43.0.0       10.20.10.10     255.255.0.0     UG        0 0          0 bond0.20
10.43.21.128    10.20.10.10     255.255.255.192 UG        0 0          0 bond0.20
10.43.67.128    10.20.10.15     255.255.255.192 UG        0 0          0 bond0.20
10.43.74.128    10.20.10.17     255.255.255.192 UG        0 0          0 bond0.20
10.43.94.192    10.20.10.11     255.255.255.192 UG        0 0          0 bond0.20
10.43.106.163   10.20.10.15     255.255.255.255 UGH       0 0          0 bond0.20
10.43.117.90    10.20.10.12     255.255.255.255 UGH       0 0          0 bond0.20
10.43.129.214   10.20.10.15     255.255.255.255 UGH       0 0          0 bond0.20
10.43.144.128   10.20.10.12     255.255.255.192 UG        0 0          0 bond0.20
10.43.192.64    10.20.10.16     255.255.255.192 UG        0 0          0 bond0.20
10.45.0.0       10.20.10.10     255.255.0.0     UG        0 0          0 bond0.20
10.45.100.100   10.20.10.15     255.255.255.255 UGH       0 0          0 bond0.20
```

It appears I'll need to flip from LoadBalancerIP to NodePort per requirements here:

>Services must be configured with the correct service type (“Cluster” or “Local”) for your implementation. For externalTrafficPolicy: Local, the service must be type LoadBalancer or NodePort.

**Comment by gclawes** [comment]
You had to manually specify the externalIP when you configured the service?  Will it work if the metallb controller assigns an IP?

**Comment by carpenike** [comment]
> You had to manually specify the externalIP when you configured the service?  Will it work if the metallb controller assigns an IP?

Manually in the sense that it didn't come out of a pool and I defined it inside of the helm deployment yes.

The BPG portion won't work with the load balancer from MetalLB which is what I was testing. I think you'd get asynchronous routes if attempting to use both IPs at the same time.

**Comment by ryholt** [comment]
Manually in the sense that I needed to specify the IP in the helm deployment and it didn’t come out of a pool.

Leaving metallb in the mix wouldn’t solve the issue with BGP, and it may have asynchronous routing problems.

**Comment by Elegant996** [comment]
Works well so long as you don't need `externalTrafficPolicy: Local` per your notes. Just wanted to re-iterate that. Thanks!

**Comment by carpenike** [comment]
Ran into an SNAT problem that requires a feature gate config for Kube proxy in 1.18. Opened a ticket for k3s -- https://github.com/rancher/k3s/issues/2090

**Comment by Elegant996** [comment]
Also up for Cherry Pick in 1.17, see kubernetes/kubernetes#90536. This would seemingly remove the need for a bare metal LB or am I missing something?

Does TCP and UDP still function correctly if you assign the same IP to two services? MetalLB offers this through `metallb.universe.tf/allow-shared-ip: "true"` presently but this would seemingly remove LBs.

**Comment by champtar** [comment]
> Also up for Cherry Pick in 1.17, see [kubernetes/kubernetes#90536](https://github.com/kubernetes/kubernetes/pull/90536). This would seemingly remove the need for a bare metal LB or am I missing something?

ExternalIPs configure the K8S dataplane, but you still need something to tell the network to send the packets to the node (ARP, BGP, OSPF, ...)

**Comment by carpenike** [comment]
Seems to remove the need for a LB altogether. I haven't had any issue with tcp/udp port sharing.

**Comment by Elegant996** [comment]
Not quite, I found out that the LB should be used for service health checks as External IPs do not perform them. Instead it just throws you at one of the endpoints without checking.

That being said, when I read this over I realized something. The point of the MetalLB controller is to create the LoadBalancer IP. So if we can route to the service through an External IP, then the controller can route to it as well making it accessible through the LoadBalancer IP.

Give this a go, install MetalLB but completely remove the speaker daemonset. You should still be able to route your ingress controller on the LoadBalancer IP without issue.

Forget BGP and with MetalLB, let Calico handle it and just use MetalLB for your LoadBalancer IP :) This completely removes the need for peering and accomplishes the same goals. Perhaps this is sufficient to close the issue?

**Comment by adamdunstan** [comment]
@Elegant996  I dont have calico running, this behavior is dependent upon kubeproxy, are you using kubeproxy from calico, iptables or ipvs?  Its correct that metallb just 'attracts" traffic.  The controller does nothing more that allocate ip addresses, if you run it without a speaker, the external IP address will still get allocated.  In addition to the speaker looking for it, kubeproxy is also.  In iptables mode a filter is added to pre-routing so traffic from that address can be forwarded, in ipvs mode its added to the ipvs interface.  Not sure how calico is adding to its routing table, but I would guess that its reading the interfaces and importing the kernel routes, if so, they will be advertised by the calico router (which is bird I think).  So the behavior is more of a side effect than intended.  Should not work with kubeproxy in iptables mode, but can work in either ipvs or I guess calico's kubeproxy....

**Comment by Elegant996** [comment]
Using kube-proxy in ipvs mode. Is there any downside to this method? Seems like a good solution otherwise.

**Comment by adamdunstan** [comment]
@Elegant996  As I guessed.  With the caveat that I havent looked at how calico is configuring bird,...  I assume that its importing the interfaces from ipvs0, and advertising those routes.  If I am correct you will be getting all the routes to all of the addresses that Ipvs has attached.  This may be what you want, but this will include endpoints and kubeapi.  You may want to modify the bird configuration (I think its in a configmap) to filter only the external addresses, and it would then advertise thoses.  Make sure that ipvs is configured with the strict_arp flag.  Bit of a confusing name, just means that the IPVS interface should not answer arp requests, otherwise every node will answer for those addresses locally, which doesnt really matter for routed destinations but could cause you some confusion later on..... Hope I have been helpful.....

**Comment by Elegant996** [comment]
@adamdunstan Sounds good, FWIW the original thought actually came from this comment here regarding kube-router and MetalLB https://github.com/metallb/metallb/issues/160#issuecomment-446708166.

**Comment by fasaxc** [comment]
As of Calico v3.16, Calico now has support for setting the BGP port that we use.  That might help if you need to peer twice to the same ToR from one node. https://docs.projectcalico.org/reference/resources/bgpconfig#spec

**Comment by johananl** [comment]
Thanks for the heads up @fasaxc! There are upcoming changes to MetalLB which would allow users to specify the source IP MetalLB uses for BGP sessions. The destination port for BGP is already configurable.

**Comment by salanki** [comment]
https://github.com/projectcalico/confd/issues/301

**Comment by caseydavenport** [comment]
Another update here - in Calico v3.18, Calico will be capable of advertising LoadBalancer IPs allocated by the MetalLB controller without installing Speaker. https://github.com/projectcalico/confd/pull/422

**Comment by ivan046** [comment]
Provide some setup guide for this please!

**Comment by gautvenk** [comment]
> Another update here - in Calico v3.18, Calico will be capable of advertising LoadBalancer IPs allocated by the MetalLB controller without installing Speaker. [projectcalico/confd#422](https://github.com/projectcalico/confd/pull/422)

I was able to try Calico v3.18 (with BGP peering to a ToR switch) along with MetalLB v0.9.5 manifests. I deleted the speaker daemon set, and confirmed that external LB IPs were advertised via BIRD to the peer. Even though I dont use the speaker component, I had to specify "protocol" as layer2 or bgp in the config-map to ensure controller allocates IPs for LB. If the protocol requirement can be removed if we dont use speakers, it will look neat. Please let me know if I can file an issue for this.

**Comment by salanki** [comment]
Very happy to hear this @gautvenk. Thank you @caseydavenport for pushing this over the line.

**Comment by russellb** [comment]
> > Another update here - in Calico v3.18, Calico will be capable of advertising LoadBalancer IPs allocated by the MetalLB controller without installing Speaker. [projectcalico/confd#422](https://github.com/projectcalico/confd/pull/422)

Thanks!  Based on this, I think we can close out this old issue.  It seems ideal to me that the existing BGP daemon would handle advertisements vs trying to colocate the current MetalLB bgp speaker.

> I was able to try Calico v3.18 (with BGP peering to a ToR switch) along with MetalLB v0.9.5 manifests. I deleted the speaker daemon set, and confirmed that external LB IPs were advertised via BIRD to the peer. Even though I don't use the speaker component, I had to specify "protocol" as layer2 or bgp in the config-map to ensure controller allocates IPs for LB. If the protocol requirement can be removed in config-map (provided we don't use speakers), it will look neat. Please let me know if I can file an issue for this.
> 
> Thanks @caseydavenport and @salanki for much awaited [projectcalico/confd#422](https://github.com/projectcalico/confd/pull/422).

Please file a feature request issue for this (or a PR is event better!).  Maybe we could support `protocol: none` in the config to be explicit that no speaker will be running, and we just want the controller for IPAM.
